File: 백준/Silver/18511. 큰 수 구성하기/큰 수 구성하기.py
n, k = map(int, input().split())
num_list = list(input().rstrip().split())
answer = 0
digit = len(str(n))

def recur(depth: int, number: list):
    global digit, num_list, k, answer
    value = int(''.join(number)) if number else 0
    if answer < value <= n:
        answer = value
    if depth == digit:
        return
    
    for num in num_list:
        number.append(num)
        recur(depth + 1, number)
        number.pop()

recur(0, [])
print(answer)

File: 백준/Silver/18511. 큰 수 구성하기/큰 수 구성하기.java
import java.util.Scanner;

class Main{
    private static int answer, n, num, len;
    private static int[] arr;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String[] str = sc.nextLine().split(" ");
        len = str[0].length();
        num = Integer.parseInt(str[0]);
        n = Integer.parseInt(str[1]);
        arr = new int[n];

        for(int i=0 ; i<n ; i++){
            arr[i] = sc.nextInt();
        }
        
        answer = 0;
        recur(0, new StringBuilder());
        System.out.println(answer);
        sc.close();
    }

    private static void recur(int depth, StringBuilder sb){
        if(depth == len){
            return;
        }

        for(int i=0 ; i<n ; i++){
            sb.append(arr[i]);
            int temp = Integer.parseInt(sb.toString());
            if(temp <= num){
                answer = Math.max(answer, temp);
            }
            recur(depth + 1, sb);
            sb.deleteCharAt(depth);
        }
    }
}